keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,153 @@
+
+`%READONLY_TOKENS_REQUEST%` - Reserved (always available) `weight` of readonly requests  for particular user or group
+
+`%MUTATING_TOKENS_LIMIT%` - Limits total mutating requests `weight` across  particular user or group.
+
+`%MUTATING_TOKENS_REQUEST%` - Reserved (always available) `weight` of mutating requests  for particular user or group
+
+`%CURRENT_READONLY_TOKENS_INFLIGHT%` - Total `weight` of inflight readonly requests from the user or group
+
+`%CURRENT_MUTATING_TOKENS_INFLIGHT%` - Total `weight` of inflight mutating requests from the user or group
+
+**Example**: We have `readonly_capacity` = 1000. `readonly.request` specified for `system:masters` is 50, for user `node` is 100, so total available readonly request `tokens` for other users is 850. In other words, members of `system:masters` in any time have at least 50 `tokens`, user `node` in any time have at least 100 `tokens`.
+
+
+### Algorithm to accept/reject readonly request <a name="algorithm-to-accept-reject-readonly-request"></a>
+
+0. User $CURRENT_USER made readonly request to apiserver, pick the request `weight` as `$REQUEST_WEIGHT`
+
+1. If $CURRENT_USER has appropriate `InflightLimit` resource, pick it as `$CURRENT_INFLIGHTLIMIT_LIMIT` and `$CURRENT_INFLIGHTLIMIT_REQUEST`
+Otherwise, among groups of the `$CURRENT_USER`, pick `InflightLimit` with max `readonly.limit` as `$CURRENT_INFLIGHTLIMIT_LIMIT` and `InflightLimit` with max `readonly.request` as `$CURRENT_INFLIGHTLIMIT_REQUEST`.
+
+2. If `$CURRENT_INFLIGHTLIMIT_LIMIT.status.readonlyInflightTokens + $REQUEST_WEIGHT > $CURRENT_INFLIGHTLIMIT_LIMIT.spec.readonly.limit`,
+**REJECT** inbound request
+
+3. Else, if `$CURRENT_INFLIGHTLIMIT_REQUEST.status.readonlyInflightTokens + $REQUEST_WEIGHT <= $CURRENT_INFLIGHTLIMIT_REQUEST.spec.readonly.request`,
+ **ACCEPT** inbound request
+
+4. Else, if (sum (`readonlyInflightTokens - readonly.request`) of all `InflightLimit` objects ) + `$REQUEST_WEIGHT`
+<=  `readonly_capacity` - (sum (`readonly.request`) of all `InflightLimit` objects),
+ **ACCEPT** inbound request
+
+5. Else, **REJECT**  inbound request
+
+For mutating requests algorithm is the same.
+
+### How to determinate request weight<a name="how-to-determinate-request-weight"></a>
+
+Implement request `weight` calculation as a function F(verb_weight, related_resources_count), where verb_weight is a constant for particular verb, related_resources_count -- count of pods, events or services -- according to requested resource type and namespace.
+
+Resources counts could be stored in the map and periodically updated in background like in etcd_object_counts metrics: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd/metrics/metrics.go#L62
+
+
+## Discussionable questions <a name="discussionable-questions"></a>
+
+### How to treat namespace-scoped lists? <a name="how-to-treat-namespace-scoped-lists"></a>
```

Comment:
Lists should be charged by the page size; if page size is omitted, make them very expensive to encourage users to use the page size feature. If you do this, there's not much difference between namespaced & cluster scoped.